go/viperutil/viper.go

```diff
@@ -0,0 +1,186 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package viperutil provides additional functions for Vitess's use of viper for
+// managing configuration of various components.
+package viperutil
+
+import (
+	"fmt"
+
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+)
+
+type Option[T any] func(val *Value[T])
+
+func WithAliases[T any](aliases ...string) Option[T] {
+	return func(val *Value[T]) {
+		val.aliases = append(val.aliases, aliases...)
+	}
+}
+
+func WithDefault[T any](t T) Option[T] {
+	return func(val *Value[T]) {
+		val.hasDefault, val.value = true, t
+	}
+}
+
+func WithEnvVars[T any](envvars ...string) Option[T] {
+	return func(val *Value[T]) {
+		val.envvars = append(val.envvars, envvars...)
+	}
+}
+
+func WithFlags[T any](flags ...string) Option[T] {
+	return func(val *Value[T]) {
+		val.flagNames = append(val.flagNames, flags...)
+	}
+}
+
+type Value[T any] struct {
+	key   string
+	value T
+
+	aliases    []string
+	flagNames  []string
+	envvars    []string
+	hasDefault bool
+
+	resolve func(string) T
+	loaded  bool
+}
+
+// NewValue returns a viper-backed value with the given key, lookup function,
+// and bind options.
+func NewValue[T any](key string, getFunc func(string) T, opts ...Option[T]) *Value[T] {
+	val := &Value[T]{
+		key:     key,
+		resolve: getFunc,
+	}
+
+	for _, opt := range opts {
+		opt(val)
+	}
+
+	return val
+}
+
+// Bind binds the value to flags, aliases, and envvars, depending on the Options
+// the value was created with.
+//
+// If the passed-in viper is nil, the global viper instance is used.
+//
+// If the passed-in flag set is nil, flag binding is skipped. Otherwise, if any
+// flag names do not match the value's canonical key, they are registered as
+// additional aliases for the value's key. This function panics if a flag name
+// is specified that is not defined on the flag set.
+func (val *Value[T]) Bind(v *viper.Viper, fs *pflag.FlagSet) {
+	if v == nil {
+		v = viper.GetViper()
+	}
+
+	aliases := val.aliases
+
+	// Bind any flags, additionally aliasing flag names to the canonical key for
+	// this value.
+	if fs != nil {
+		for _, name := range val.flagNames {
+			f := fs.Lookup(name)
+			if f == nil {
+				// TODO: Don't love "configuration error" here as it might make
+				// users think _they_ made a config error, but really it is us,
+				// the Vitess authors, that have misconfigured a flag binding in
+				// the source code somewhere.
+				panic(fmt.Sprintf("configuration error: attempted to bind %s to unspecified flag %s", val.key, name))
```

Comment:
basically they can fork/edit the code, or file an issue

this _should_ always be caught in a test, provided at least one `cmd` that includes the flag is executed as a binary (e.g. we have something that spawns `./bin/vttablet` or whatever)

Reply:
ooooh i just realized this presents a problem, as it means that a viper value that has a flag associated with it (e.g. I recall there was at least one case during the flag migration where a binary didn't want to expose a flag, but it did depend on other code having the correct/initialized default value) doesn't get bound to the viper config at all.

i think we need another strategy here, because there are cases where we would want to fail loudly if we tried to bind to a flag that doesn't exist, but also, _for the exact same value_ there are cases where we wouldn't ....... 🤔

Reply:
fixed in https://github.com/vitessio/vitess/pull/11456/commits/7fac28473ce0e104b8424b8edc9eaccfabe9ab69